crates/ruff_python_formatter/src/expression/expr_slice.rs

```diff
@@ -1,24 +1,269 @@
+        // In placements.rs, we marked comment for None nodes a dangling and associated all others
+        // as leading or dangling wrt to a node. That means we either format a node and only have
+        // to handle newlines and spacing, or the node is None and we insert the corresponding
+        // slice of dangling comments
+        let comments = f.context().comments().clone();
+        let slice_dangling_comments = comments.dangling_comments(item.as_any_node_ref());
+        // Put the dangling comments (where the nodes are missing) into buckets
+        let first_colon_partition_index = slice_dangling_comments
+            .partition_point(|x| x.slice().start() < first_colon.range.start());
+        let second_colon_partition_index =
+            second_colon
+                .as_ref()
+                .map_or(slice_dangling_comments.len(), |second_colon| {
+                    slice_dangling_comments[first_colon_partition_index..]
+                        .partition_point(|x| x.slice().start() < second_colon.range.start())
+                        + first_colon_partition_index
+                });
+        // Ensure there a no dangling comments for a node if the node is present
+        debug_assert!(!(lower.is_some() && first_colon_partition_index > 0));
+        debug_assert!(
+            !(upper.is_some() && first_colon_partition_index < second_colon_partition_index)
+        );
+        debug_assert!(
+            !(step.is_some() && second_colon_partition_index < slice_dangling_comments.len())
+        );
+
+        // Handle spacing around the colon(s)
+        // https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#slices
+        let lower_simple = lower.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let upper_simple = upper.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let step_simple = step.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let all_simple = lower_simple && upper_simple && step_simple;
+
+        // lower
+        if let Some(lower) = lower {
+            lower.format().fmt(f)?;
+            if comments.has_trailing_comments(lower.as_ref()) {
+                // make sure the colon is after the comments
+                hard_line_break().fmt(f)?;
+            }
+        } else {
+            dangling_comments(&slice_dangling_comments[..first_colon_partition_index]).fmt(f)?;
+        }
+
+        // First colon
+        // The spacing after the colon depends on both the lhs and the rhs:
+        // ```
+        // e00 = x[:]
+        // e01 = x[:1]
+        // e02 = x[: a()]
+        // e10 = x[1:]
+        // e11 = x[1:1]
+        // e12 = x[1 : a()]
+        // e20 = x[a() :]
+        // e21 = x[a() : 1]
+        // e22 = x[a() : a()]
+        // e200 = "e"[a() : :]
+        // e201 = "e"[a() :: 1]
+        // e202 = "e"[a() :: a()]
+        // ```
+        if !all_simple {
+            space().fmt(f)?;
+        }
+        text(":").fmt(f)?;
```

Comment:
You don't like the `write` macro, do you ;)

Reply:
procedural code is kinda convenient :see_no_evil: